File: matlab/layerS1.m
%***********************
% SIMPLE CELL LEARNING *
%***********************

rand('state',PARAM.randState);

for i=1:PARAM.c.RFSize(1)
    for j=1:PARAM.c.RFSize(2)
        
        timedLogLn(['S1 MAP (' int2str(i) ',' int2str(j) ')']);

        load([PATH.map 'onOff.' int2str(i) '.' int2str(j) '.mat']);
        COMMON.onOffMap = onOffMap;
        clear onOffMap;

        % LEARNING
        if true % learn S1 selectivity
            %             [center, nCouldHaveFired, nFired] = competitiveHebbianCenterLearning(length(COMMON.onOffMap),PARAM.s.nIter,@getOnOffMap,false,[],[],PARAM.recFilter,PARAM.s,PARAM.cropPos);
            %             [center, nCouldHaveFired, nFired] = competitiveHebbianCenterLearning(length(COMMON.onOffMap),PARAM.s.nIter,@getOnOffMap,false,[],[],PARAM.recFilter,PARAM.s);
%                         [center, nCouldHaveFired, nFired, w ] = antiHebbianLearning(length(COMMON.onOffMap),PARAM.s.nIter,@getOnOffMap,[],[],[],PARAM.recFilter,PARAM.s);
            [center] = learnUnlearn(length(COMMON.onOffMap),PARAM.s.nIter,@getOnOffMap,[],[],PARAM.recFilter,PARAM.s);
%                                 rec(center,PARAM.recFilter,'')
%                                 dp = analyzeCenter(center,'');
%                                 error('stop')

            % free memory
            if isfield(center,'evolDist')
                center = rmfield(center,'evolDist');
            end
            if isfield(center,'evolThetaOriginal')
                center = rmfield(center,'evolThetaOriginal');
            end
            if isfield(center,'evolDeltaTheta')
                center = rmfield(center,'evolDeltaTheta');
            end
            if isfield(center,'evolTheta')
                center = rmfield(center,'evolTheta');
            end
            if isfield(center,'evolBand')
                center = rmfield(center,'evolBand');
            end
            clear nFired nCouldHaveFired

        else % use hard wired gabor filter

            [fSiz,filters,c1OL,numSimpleFilters] = init_gabor(180/(PARAM.s.n/2)*[0:PARAM.s.n/2-1], PARAM.s.RFSize(1), .5, .5);
            for r=1:(PARAM.s.n/2)

                filter(:,:,r) = reshape(filters(:,r),[PARAM.s.RFSize]);
                center(r).patch(:,:,1) = max(0,filter(:,:,r));
                center(r).patch(:,:,2) = -min(0,filter(:,:,r));
                center(r+PARAM.s.n/2).patch(:,:,1) = -min(0,filter(:,:,r));
                center(r+PARAM.s.n/2).patch(:,:,2) = max(0,filter(:,:,r));


                %                 figure
                %                 imagesc(filter(:,:,r))
                %                 colormap(gray)
                %                 disp(filter(:,:,r))
                %                 pause
                %         filter(:,:,r) = filter(end:-1:1,end:-1:1,r); % flip in order to use conv2
            end
        end
        % format i_s x j_s x i_onOff x j_onOfff x iCenter x 2
        COMMON.center(i,j,:,:,:,:) = reshape([center.patch],[PARAM.s.RFSize length(center) 2]);

%         recAll(COMMON.center,PARAM.recFilter,[ 'All S1' PARAM.compName])
%         error('stop')

%        error('stop')

        % RECORD RESPONSE
        for block=1:ceil( length(COMMON.onOffMap) / PARAM.c.block )
            name = ['s1.' sprintf('%02d',block) '.' PARAM.s.type '.mat'];
            if i==1 && j==1
                s1Map = zeros([PARAM.c.RFSize PARAM.s.n PARAM.c.block]);
            else
                load([PATH.map name])
            end
            range = 1+(block-1)*PARAM.c.block:min(length(COMMON.onOffMap),block*PARAM.c.block);
%             s1Map(i,j,:,1:length(range)) = s1(:,range);% format i_s x j_s x iCenter x iMap
            s1Map(i,j,:,1:length(range)) = recordResponse(range,@getOnOffMap,false,COMMON.center(i,j,:,:,:,:));% format iCenter x iMap
            save([PATH.map name],'s1Map');
            clear s1Map
        end
        COMMON.nMap = length(COMMON.onOffMap);
        COMMON = rmfield(COMMON,'onOffMap');
%         clear s1
    end
end
save([PATH.map 'COMMON.' PARAM.compName '.' PARAM.s.type '.mat'],'COMMON','PARAM')
currentPath = pwd;
cd(PATH.map)
copyfile(['COMMON.' PARAM.compName '.' PARAM.s.type '.mat'],['../COMMON.mat'])
cd(currentPath); clear currentPath;

recAll(COMMON.center,PARAM.recFilter,[ 'All S1' PARAM.compName])

% error('stop')

File: python/layer_s1.py
from config import *
from learn_unlearn import *
from numpy import *
import pickle
from tools import *


def record_response(range_var, weight_sharig_mode, center, common):
    '''
    Timothee Masquelier timothee.masquelier@alum.mit.edu August 2006
    n_map: number of input maps
    get_map: handle to a function i -> map
    weight_sharing_mode: if true convolution mode + max response, if false simple local dot product
    center: structure that must contains at least center.patch
    '''
    n_center = shape(center)[2]
    
    patch = []
    for c_ind in range(shape(center)[2]):
        tmp = center[:, :, c_ind, :]
        patch.append(tmp.flatten(1))
          
    response = zeros((n_center, len(range_var)), float64)
    print 'Recording responces...'
    if weight_sharig_mode:
        raise NameError('this does not work yet')
    else:
        for m_ind in range_var:
            img = get_onoff_map(m_ind, common)
            img = img / linalg.norm(img.flatten(1))
            
            for cent_ind in range(n_center): 
                response[cent_ind, m_ind - range_var[0]] = dot(img.flatten(1),
                                                               patch[cent_ind])
      
    return response
    


def layer_s1():

    common = {}
    common['center'] = zeros((params['c']['rf_size'][0], 
                            params['c']['rf_size'][1],
                            params['s']['rf_size'][0],
                            params['s']['rf_size'][1], params['s']['n'], 2))
    for i_ind in range(params['c']['rf_size'][0]):
        for j_ind in range(params['c']['rf_size'][1]):
            print 'S1 MAP (' + str(i_ind) + ', ' + str(j_ind) + ')'
            
            common['on_off_map'] = load(paths['map'] + 'onoff.' + str(i_ind) + \
                                        '.' + str(j_ind) + '.npy')
            #learning
            center = learn_unlearn(len(common['on_off_map']),
                                   params['s']['n_iter'], [], [],
                                   params['rec_filter'], params['s'],
                                   common)
           
    	    tmp_cent = zeros((params['s']['rf_size'][0],
                             params['s']['rf_size'][1], len(center), 2))
            for cent_ind in range(len(center)):
                tmp_cent[:, :, cent_ind, :] = center[cent_ind]['patch']
           
            
            common['center'][i_ind, j_ind, :, :, :, :] = tmp_cent
            
            #record responce
            for block in range(ceil(float(len(common['on_off_map'])) / \
                               params['c']['block'])):
                name = 's1.' + str(block) + '.' + params['s']['type'] + '.npy'
           
                if i_ind == 0 and j_ind == 0:
                    s1_map = zeros((params['c']['rf_size'][0],
                                   params['c']['rf_size'][1], params['s']['n'],
                                   params['c']['block']))
                else:
                    s1_map = load(paths['map'] + name)

                range_var = range(block * params['c']['block'],
                                  minimum(len(common['on_off_map']),
                                  (block + 1) * params['c']['block']))
                s1_map[i_ind, j_ind, :, 0:len(range_var)] = \
                    record_response(range_var, False,
                                    common['center'][i_ind, j_ind, :, :, :, :],
                                    common)
                save(paths['map'] + name, s1_map)

            common['n_map'] = len(common['on_off_map'])
            #clear s1
            del common['on_off_map'] 

    cfile = open(paths['map'] + 'common.' + params['comp_name'] + '.' + \
                 params['s']['type'] + '.pck', 'w')
    pickle.dump(common, cfile)
    cfile.close()


if __name__ == "__main__":
    #params['test_mode'] = True
    layer_s1()

